File: src/main/java/net/nodus/database/analytics/AnalyticsEventRepository.java
package net.nodus.database.analytics;

import java.time.Instant;
import java.util.Optional;
import org.springframework.data.mongodb.repository.MongoRepository;

public interface AnalyticsEventRepository extends MongoRepository<AnalyticsEvent, String> {

    Optional<AnalyticsEvent> findFirstBySiteIdAndSessionIdAndEventTypeAndEventTimeLessThanOrderByEventTimeDesc(
        String siteId,
        String sessionId,
        String eventType,
        Instant eventTime
    );
}

File: src/main/kotlin/net/nodus/analytics/AnalyticsEventRepository.kt
package net.nodus.analytics

import org.springframework.data.mongodb.repository.MongoRepository
import java.time.Instant

interface AnalyticsEventRepository : MongoRepository<AnalyticsEvent, String> {
    fun findFirstBySiteIdAndSessionIdAndEventTypeAndEventTimeLessThanOrderByEventTimeDesc(
        siteId: String,
        sessionId: String,
        eventType: String,
        eventTime: Instant,
    ): AnalyticsEvent?
}